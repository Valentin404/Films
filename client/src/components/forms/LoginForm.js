import React, {useState} from "react"
import {Link} from "react-router-dom"
import FormMessage from "./FormMessage"
import setFormObject from "./FormUtils"

const initialData = {
  email: "",
  password: "",
}

const LoginForm = props => {
  const [form, setForm] = useState(initialData)
  const [loading, setLoading] = useState(false)
  const [errors, setErrors] = useState({})

  const handleSubmit = e => {
    e.preventDefault()
    const errors = validate(form)
    setErrors(errors)
    if (Object.keys(errors).length === 0) {
      setLoading(true)
      props.submit(form).catch(error => {
        setErrors(error.response.data.errors)
        setLoading(false)
      })
    }
  }

  const validate = (data) => {
    const errors = {}
    if (!data.email) errors.email = "Email cannot be blank"
    if (!data.password) errors.password = "Password cannot be blank"

    return errors
  }

  const cls = loading ? "ui form loading" : "ui form"

    return (
      <form className={cls} onSubmit={handleSubmit}>
        <div className={errors.email ? "error field" : "field"}>
          <label htmlFor={'email'}>Email</label>
          <input
            type="email"
            name="email"
            id="email"
            placeholder="Email"
            value={form.email}
            onChange={setFormObject(form, setForm)}
          />
          {errors.email && <FormMessage>{errors.email}</FormMessage>}
        </div>

        <div className={errors.password ? "error field" : "field"}>
          <label htmlFor={'password'}>Password</label>
          <input
            type="text"
            name="password"
            id="password"
            placeholder="password"
            value={form.password}
            onChange={setFormObject(form, setForm)}
          />
          {errors.password && <FormMessage>{errors.password}</FormMessage>}
        </div>
        <div className="ui fluid buttons">
          <button className="ui button primary" data-testid='login-button'>Login</button>

          <div className="or" />

          <Link to="/" className="ui button">
            Cancel
          </Link>
        </div>
      </form>
    )
}

export default LoginForm
