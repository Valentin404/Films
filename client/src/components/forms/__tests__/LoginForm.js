import React from "react"
import {MemoryRouter as Router} from "react-router-dom"
import user from "@testing-library/user-event"
import {render, fireEvent} from "@testing-library/react"
import LoginForm from "../LoginForm"

const submit = jest.fn(() => Promise.resolve())

const data = {
    email: "test@mail.com",
    password: "mypass",
}

test("should render correctly", () => {
    const {getByLabelText, getByTestId, getByRole, queryByRole} = render(<Router><LoginForm submit={submit}/></Router>)

    const emailEl = getByLabelText(/email/i)
    const passwordEl = getByLabelText(/password/i)
    const btn = getByTestId('login-button');

    user.type(emailEl, 'not correct email');
    user.type(passwordEl, 'correct password');

    fireEvent.click(btn)

    expect(queryByRole('alert')).toBeNull()
})
