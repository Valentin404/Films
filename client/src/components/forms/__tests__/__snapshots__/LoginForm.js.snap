// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render correctly 1`] = `
<form
  class="ui form"
>
  <div
    class="field"
  >
    <label
      for="email"
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      placeholder="Email"
      type="email"
      value=""
    />
    <div>
      <span
        style="color: rgb(154, 63, 56);"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label
      for="password"
    >
      Password
    </label>
    <input
      id="password"
      name="password"
      placeholder="password"
      type="text"
      value=""
    />
    <div>
      <span
        style="color: rgb(154, 63, 56);"
      />
    </div>
  </div>
  <div
    class="ui fluid buttons"
  >
    <button
      class="ui button primary"
      data-testid="login-button"
    >
      Login
    </button>
    <div
      class="or"
    />
    <a
      class="ui button"
      href="/"
    >
      Cancel
    </a>
  </div>
</form>
`;
